File: demo-images/generate_samples.py
#!/usr/bin/env python3
"""
Generate sample images with variations for perceptual hashing demo.
These images will demonstrate similar photos that perceptual hashing can detect.
"""

from PIL import Image, ImageEnhance, ImageFilter
import os

def create_base_image(size=(256, 256), bg_color=(135, 206, 235), sun_color=(255, 215, 0)):
    """Create a base sunset image"""
    img = Image.new('RGB', size, bg_color)

    # Create a gradient for sunset
    for y in range(size[1]):
        progress = y / size[1]
        r = int(bg_color[0] + (255 - bg_color[0]) * progress)
        g = int(bg_color[1] + (100 - bg_color[1]) * progress)
        b = int(bg_color[2] + (50 - bg_color[2]) * progress)
        for x in range(size[0]):
            img.putpixel((x, y), (r, g, b))

    # Add sun
    sun_radius = size[0] // 8
    sun_x = size[0] // 2
    sun_y = size[1] // 3

    for y in range(size[1]):
        for x in range(size[0]):
            dist = ((x - sun_x) ** 2 + (y - sun_y) ** 2) ** 0.5
            if dist < sun_radius:
                img.putpixel((x, y), sun_color)

    return img

def create_base_cat_image(size=(256, 256), bg_color=(240, 248, 255)):
    """Create a simple cat silhouette image"""
    img = Image.new('RGB', size, bg_color)

    # Cat body (simple shapes)
    cat_color = (100, 100, 100)

    # Body (ellipse)
    for y in range(size[1]):
        for x in range(size[0]):
            # Body
            cx, cy = size[0] // 2, size[1] // 2 + 20
            if ((x - cx) ** 2) / 40 ** 2 + ((y - cy) ** 2) / 50 ** 2 <= 1:
                img.putpixel((x, y), cat_color)

            # Head
            hx, hy = size[0] // 2, size[1] // 3
            if ((x - hx) ** 2) / 35 ** 2 + ((y - hy) ** 2) / 30 ** 2 <= 1:
                img.putpixel((x, y), cat_color)

            # Left ear
            ex1, ey1 = hx - 20, hy - 15
            ex2, ey2 = hx - 10, hy - 35
            ex3, ey3 = hx - 5, hy - 15
            # Simple triangle check
            if point_in_triangle(x, y, ex1, ey1, ex2, ey2, ex3, ey3):
                img.putpixel((x, y), cat_color)

            # Right ear
            ex1, ey1 = hx + 20, hy - 15
            ex2, ey2 = hx + 10, ey2
            ex3, ey3 = hx + 5, ey3
            if point_in_triangle(x, y, ex1, ey1, ex2, ey2, ex3, ey3):
                img.putpixel((x, y), cat_color)

    return img

def point_in_triangle(px, py, x1, y1, x2, y2, x3, y3):
    """Check if point is in triangle using barycentric coordinates"""
    def area(ax, ay, bx, by, cx, cy):
        return abs((ax * (by - cy) + bx * (cy - ay) + cx * (ay - by)) / 2.0)

    A = area(x1, y1, x2, y2, x3, y3)
    A1 = area(px, py, x2, y2, x3, y3)
    A2 = area(x1, y1, px, py, x3, y3)
    A3 = area(x1, y1, x2, y2, px, py)

    return abs(A - (A1 + A2 + A3)) < 0.1

def create_variations(base_img, output_dir, prefix, count=5):
    """Create variations of base image"""
    variations = []

    for i in range(count):
        img = base_img.copy()

        # Apply different transformations based on index
        if i == 1:
            # Slightly brighter
            enhancer = ImageEnhance.Brightness(img)
            img = enhancer.enhance(1.15)
            filename = f"{prefix}_brightened.jpg"
        elif i == 2:
            # Slightly more contrast
            enhancer = ImageEnhance.Contrast(img)
            img = enhancer.enhance(1.2)
            filename = f"{prefix}_contrast.jpg"
        elif i == 3:
            # Slight blur
            img = img.filter(ImageFilter.GaussianBlur(radius=0.5))
            filename = f"{prefix}_blurred.jpg"
        elif i == 4:
            # Saturation boost (Instagram-like)
            enhancer = ImageEnhance.Color(img)
            img = enhancer.enhance(1.3)
            filename = f"{prefix}_saturated.jpg"
        else:
            # Original
            filename = f"{prefix}_original.jpg"

        filepath = os.path.join(output_dir, filename)
        img.save(filepath, 'JPEG', quality=90)
        variations.append(filepath)
        print(f"Created: {filename}")

    return variations

def main():
    output_dir = '/home/ubuntu/file-deduplicator/demo-images/samples'
    os.makedirs(output_dir, exist_ok=True)

    print("Generating sample images for perceptual hashing demo...")
    print("=" * 60)

    # Create sunset variations
    print("\n🌅 Creating sunset image variations...")
    sunset_base = create_base_image()
    sunset_files = create_variations(sunset_base, output_dir, 'sunset')

    # Create cat variations
    print("\n🐱 Creating cat image variations...")
    cat_base = create_base_cat_image()
    cat_files = create_variations(cat_base, output_dir, 'cat')

    print("\n" + "=" * 60)
    print(f"✅ Generated {len(sunset_files) + len(cat_files)} sample images")
    print(f"📁 Location: {output_dir}")
    print("\nThese images have slight variations that perceptual hashing can detect:")
    print("  - Original")
    print("  - Brightened (+15%)")
    print("  - Contrast boosted (+20%)")
    print("  - Slightly blurred")
    print("  - Saturated (+30%)")

if __name__ == '__main__':
    main()

File: demo-images/generate_samples.go
package main

import (
	"image"
	"image/color"
	"image/jpeg"
	"os"
)

func createSunsetImage(path string, brightness float64) error {
	img := image.NewRGBA(image.Rect(0, 0, 256, 256))

	// Create sunset gradient
	for y := 0; y < 256; y++ {
		progress := float64(y) / 256.0
		r := uint8(135 + (255-135)*progress*brightness)
		g := uint8(206 + (100-206)*progress*brightness)
		b := uint8(235 + (50-235)*progress*brightness)
		for x := 0; x < 256; x++ {
			img.Set(x, y, color.RGBA{r, g, b, 255})
		}
	}

	// Add sun
	sunX, sunY := 128, 85
	sunRadius := 32
	for y := 0; y < 256; y++ {
		for x := 0; x < 256; x++ {
			dx := x - sunX
			dy := y - sunY
			if dx*dx+dy*dy < sunRadius*sunRadius {
				img.Set(x, y, color.RGBA{255, 215, 0, 255})
			}
		}
	}

	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	return jpeg.Encode(f, img, &jpeg.Options{Quality: 90})
}

func createCatImage(path string) error {
	img := image.NewRGBA(image.Rect(0, 0, 256, 256))

	// Background
	for y := 0; y < 256; y++ {
		for x := 0; x < 256; x++ {
			img.Set(x, y, color.RGBA{240, 248, 255, 255})
		}
	}

	// Simple cat silhouette
	catColor := color.RGBA{100, 100, 100, 255}

	// Body
	for y := 140; y < 200; y++ {
		for x := 88; x < 168; x++ {
			cx, cy := 128, 170
			dx, dy := float64(x-cx)/40, float64(y-cy)/30
			if dx*dx+dy*dy <= 1.0 {
				img.Set(x, y, catColor)
			}
		}
	}

	// Head
	for y := 70; y < 130; y++ {
		for x := 93; x < 163; x++ {
			cx, cy := 128, 100
			dx, dy := float64(x-cx)/35, float64(y-cy)/30
			if dx*dx+dy*dy <= 1.0 {
				img.Set(x, y, catColor)
			}
		}
	}

	// Left ear
	for y := 65; y < 85; y++ {
		for x := 93; x < 113; x++ {
			if x+y > 158 && x+3*y > 300 && 3*x+y < 420 {
				img.Set(x, y, catColor)
			}
		}
	}

	// Right ear
	for y := 65; y < 85; y++ {
		for x := 143; x < 163; x++ {
			if 508-x+y > 158 && 508-x+3*y > 300 && 768-3*x+y < 420 {
				img.Set(x, y, catColor)
			}
		}
	}

	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	return jpeg.Encode(f, img, &jpeg.Options{Quality: 90})
}

func main() {
	os.MkdirAll("/home/ubuntu/file-deduplicator/demo-images/samples", 0755)

	files := []struct {
		path       string
		brightness float64
	}{
		{"samples/sunset_original.jpg", 1.0},
		{"samples/sunset_bright.jpg", 1.15},
		{"samples/sunset_dark.jpg", 0.85},
		{"samples/cat_original.jpg", 1.0},
		{"samples/cat_copy1.jpg", 1.0},
		{"samples/cat_copy2.jpg", 1.0},
	}

	for _, f := range files {
		if len(f.path) > 10 && f.path[:10] == "samples/sun" {
			err := createSunsetImage("/home/ubuntu/file-deduplicator/demo-images/"+f.path, f.brightness)
			if err != nil {
				panic(err)
			}
			println("Created:", f.path)
		} else if len(f.path) > 10 && f.path[:10] == "samples/cat" {
			err := createCatImage("/home/ubuntu/file-deduplicator/demo-images/"+f.path)
			if err != nil {
				panic(err)
			}
			println("Created:", f.path)
		}
	}
}
